src/Features/Core/Portable/AddPackage/ParentInstallPackageCodeAction.cs

```diff
@@ -0,0 +1,129 @@
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.AddPackage;
+
+/// <summary>
+/// Data used to create the 'Install Nuget Package' top level code-action. It will have children to 'Install Latest',
+/// 'Install Version 'X' ..., and 'Install with package manager'.
+/// </summary>
+/// <param name="packageSource">The nuget source to use.  Currently this is only <see
+/// cref="PackageSourceHelper.NugetOrgSourceName"/> ("::nuget::")</param>
+/// <param name="packageName">The name of the package to install.</param>
+/// <param name="packageVersionOpt">A preferred version if known.</param>
+/// <param name="textChanges">Additional text changes to make to the <see cref="Document"/>.  Generally, this would be
+/// the import to add if not present.</param>
+internal readonly struct InstallPackageData(string packageSource, string packageName, string packageVersionOpt, ImmutableArray<TextChange> textChanges)
+{
+    public readonly string PackageSource = packageSource;
+    public readonly string PackageName = packageName;
+    public readonly string PackageVersionOpt = packageVersionOpt;
+
+    public readonly ImmutableArray<TextChange> TextChanges = textChanges;
+}
+
+/// <summary>
+/// This is the top level 'Install Nuget Package' code action we show in 
+/// the lightbulb.  It will have children to 'Install Latest', 
+/// 'Install Version 'X' ..., and 'Install with package manager'.
+/// </summary>
+internal sealed class ParentInstallPackageCodeAction : CodeAction.CodeActionWithNestedActions
+{
+    /// <summary>
+    /// This code action only works by installing a package.  As such, it requires a non document change (and is
+    /// thus restricted in which hosts it can run).
+    /// </summary>
+    public override ImmutableArray<string> Tags => RequiresNonDocumentChangeTags;
+
+    /// <summary>
+    /// Even though we have child actions, we mark ourselves as explicitly non-inlinable.
+    /// We want to the experience of having the top level item the user has to see and
+    /// navigate through, and we don't want our child items confusingly being added to the
+    /// top level light-bulb where it's not clear what effect they would have if invoked.
+    /// </summary>
+    public ParentInstallPackageCodeAction(
+        Document document,
+        InstallPackageData fixData,
+        IPackageInstallerService installerService)
+        : base(string.Format(FeaturesResources.Install_package_0, fixData.PackageName),
+               CreateNestedActions(document, fixData, installerService),
+               isInlinable: false,
+               priority: CodeActionPriority.Low) // Adding a nuget reference is lower priority than other fixes..
+    {
+    }
+
+    public static CodeAction? TryCreateCodeAction(
+        Document document,
+        InstallPackageData fixData,
+        IPackageInstallerService? installerService)
```

Comment:
this is teh factory method that we use ourselves *and* which the ASP EA layer will call into.